blog/2024-09-12-building-the-opentofu-registry.md

```diff
@@ -0,0 +1,129 @@
+
+### The OpenTofu Docs API (and how you can use it)
+
+Doing all this work on the backend and splitting the display logic from the data also had an unintended but very welcome side effect: we were able to offer [an API for provider and module documentation](https://api.opentofu.org), which would come in handy as just a few weeks later [Jetbrains requested just such an API for their OpenTofu integration](https://youtrack.jetbrains.com/issue/IJPL-157423/Support-for-OpenTofu-in-Terraform-and-HCL-Plugin#focus=Comments-27-10245678.0-0).
+
+The backend produces the dataset in this format and is easily runnable locally from the [registry-ui](https://github.com/opentofu/registry-ui) repository, while the [public API](https://api.opentofu.org) is available for anyone to build integrations with. We even made sure to include the correct [CORS headers](https://en.wikipedia.org/wiki/Cross-Origin_Resource_Sharing) if you would like to build a browser-only integration. If you build something cool with it, let us know!
+
+## Search
+
+As we were building the backend, one issue was constantly on our minds: how do we make such a vast data set searchable? We hoped that it would be possible to simply generate a search index and let the client JavaScript deal with the entire search problem. As we were looking at [lunr.js](https://lunrjs.com/), a robust and mature library for search, it quickly became apparent that this path would be entirely unfeasible. Even with a limited data set, the download size of the search index quickly ballooned to several hundred megabytes, which is not exactly ideal for a snappy search functionality not to mention really upsetting to people with data caps.
+
+Not wanting to run our own database server or involve any more service dependencies, we looked at [Cloudflare's D1](https://developers.cloudflare.com/d1/), a SQLite database service and a [worker](https://workers.cloudflare.com/) to handle search queries. While it initially looked promising, we discovered that we would have to submit all updates in a single HTTP request in order to run them into a transaction. 
+
+It would have been possible to work around it and still perform atomic search index updates, but we ended up using [Neon](https://neon.tech), a database-as-a-service offering, instead. They don't explicitly sponsor OpenTofu, but their free tier was comfortably enough for our search index and the next tier was also quite affordable. Their tight integration with Cloudflare was also a very welcome addition.
```

Comment:
```suggestion
It would have been possible to work around it and still perform atomic search index updates, but we ended up using [Neon](https://neon.tech), a database-as-a-service offering, instead. They don't explicitly sponsor OpenTofu, but their free tier was comfortable enough for our search index and the next tier was also quite affordable. Their tight integration with Cloudflare was also a very welcome addition.
```